<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Python programming I</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark-2b3e328b71be8d25427581baeb23079b.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" class="quarto-color-scheme-extra" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-b464d2b13067cecfcf6852c95fd93919.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark-538525e415c347f693b47e32e687dffb.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="site_libs/bootstrap/bootstrap-b464d2b13067cecfcf6852c95fd93919.min.css" rel="stylesheet" append-hash="true" class="quarto-color-scheme-extra" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="floating nav-fixed quarto-light"><script id="quarto-html-before-body" type="application/javascript">
    const toggleBodyColorMode = (bsSheetEl) => {
      const mode = bsSheetEl.getAttribute("data-mode");
      const bodyEl = window.document.querySelector("body");
      if (mode === "dark") {
        bodyEl.classList.add("quarto-dark");
        bodyEl.classList.remove("quarto-light");
      } else {
        bodyEl.classList.add("quarto-light");
        bodyEl.classList.remove("quarto-dark");
      }
    }
    const toggleBodyColorPrimary = () => {
      const bsSheetEl = window.document.querySelector("link#quarto-bootstrap:not([rel=disabled-stylesheet])");
      if (bsSheetEl) {
        toggleBodyColorMode(bsSheetEl);
      }
    }
    const setColorSchemeToggle = (alternate) => {
      const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
      for (let i=0; i < toggles.length; i++) {
        const toggle = toggles[i];
        if (toggle) {
          if (alternate) {
            toggle.classList.add("alternate");
          } else {
            toggle.classList.remove("alternate");
          }
        }
      }
    };
    const toggleColorMode = (alternate) => {
      // Switch the stylesheets
      const primaryStylesheets = window.document.querySelectorAll('link.quarto-color-scheme:not(.quarto-color-alternate)');
      const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
      manageTransitions('#quarto-margin-sidebar .nav-link', false);
      if (alternate) {
        // note: dark is layered on light, we don't disable primary!
        enableStylesheet(alternateStylesheets);
        for (const sheetNode of alternateStylesheets) {
          if (sheetNode.id === "quarto-bootstrap") {
            toggleBodyColorMode(sheetNode);
          }
        }
      } else {
        disableStylesheet(alternateStylesheets);
        enableStylesheet(primaryStylesheets)
        toggleBodyColorPrimary();
      }
      manageTransitions('#quarto-margin-sidebar .nav-link', true);
      // Switch the toggles
      setColorSchemeToggle(alternate)
      // Hack to workaround the fact that safari doesn't
      // properly recolor the scrollbar when toggling (#1455)
      if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
        manageTransitions("body", false);
        window.scrollTo(0, 1);
        setTimeout(() => {
          window.scrollTo(0, 0);
          manageTransitions("body", true);
        }, 40);
      }
    }
    const disableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        stylesheet.rel = 'disabled-stylesheet';
      }
    }
    const enableStylesheet = (stylesheets) => {
      for (let i=0; i < stylesheets.length; i++) {
        const stylesheet = stylesheets[i];
        if(stylesheet.rel !== 'stylesheet') { // for Chrome, which will still FOUC without this check
          stylesheet.rel = 'stylesheet';
        }
      }
    }
    const manageTransitions = (selector, allowTransitions) => {
      const els = window.document.querySelectorAll(selector);
      for (let i=0; i < els.length; i++) {
        const el = els[i];
        if (allowTransitions) {
          el.classList.remove('notransition');
        } else {
          el.classList.add('notransition');
        }
      }
    }
    const isFileUrl = () => {
      return window.location.protocol === 'file:';
    }
    const hasAlternateSentinel = () => {
      let styleSentinel = getColorSchemeSentinel();
      if (styleSentinel !== null) {
        return styleSentinel === "alternate";
      } else {
        return false;
      }
    }
    const setStyleSentinel = (alternate) => {
      const value = alternate ? "alternate" : "default";
      if (!isFileUrl()) {
        window.localStorage.setItem("quarto-color-scheme", value);
      } else {
        localAlternateSentinel = value;
      }
    }
    const getColorSchemeSentinel = () => {
      if (!isFileUrl()) {
        const storageValue = window.localStorage.getItem("quarto-color-scheme");
        return storageValue != null ? storageValue : localAlternateSentinel;
      } else {
        return localAlternateSentinel;
      }
    }
    const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
      const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
      const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
      let newTheme = '';
      if(authorPrefersDark) {
        newTheme = isAlternate ? baseTheme : alternateTheme;
      } else {
        newTheme = isAlternate ? alternateTheme : baseTheme;
      }
      const changeGiscusTheme = () => {
        // From: https://github.com/giscus/giscus/issues/336
        const sendMessage = (message) => {
          const iframe = document.querySelector('iframe.giscus-frame');
          if (!iframe) return;
          iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
        }
        sendMessage({
          setConfig: {
            theme: newTheme
          }
        });
      }
      const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
      if (isGiscussLoaded) {
        changeGiscusTheme();
      }
    };
    const authorPrefersDark = false;
    const darkModeDefault = authorPrefersDark;
      document.querySelector('link#quarto-text-highlighting-styles.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
      document.querySelector('link#quarto-bootstrap.quarto-color-scheme-extra').rel = 'disabled-stylesheet';
    let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
    // Dark / light mode switch
    window.quartoToggleColorScheme = () => {
      // Read the current dark / light value
      let toAlternate = !hasAlternateSentinel();
      toggleColorMode(toAlternate);
      setStyleSentinel(toAlternate);
      toggleGiscusIfUsed(toAlternate, darkModeDefault);
      window.dispatchEvent(new Event('resize'));
    };
    // Switch to dark mode if need be
    if (hasAlternateSentinel()) {
      toggleColorMode(true);
    } else {
      toggleColorMode(false);
    }
  </script>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Python programming I</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul class="collapse">
  <li><a href="#practice-makes-perfect" id="toc-practice-makes-perfect" class="nav-link" data-scroll-target="#practice-makes-perfect">Practice makes perfect</a></li>
  <li><a href="#why-use-python" id="toc-why-use-python" class="nav-link" data-scroll-target="#why-use-python">Why use Python?</a></li>
  <li><a href="#which-other-languages-do-bioinformaticians-use" id="toc-which-other-languages-do-bioinformaticians-use" class="nav-link" data-scroll-target="#which-other-languages-do-bioinformaticians-use">Which other languages do bioinformaticians use?</a></li>
  <li><a href="#a-small-warning" id="toc-a-small-warning" class="nav-link" data-scroll-target="#a-small-warning">A small warning</a></li>
  <li><a href="#how-will-this-course-work" id="toc-how-will-this-course-work" class="nav-link" data-scroll-target="#how-will-this-course-work">How will this course work?</a></li>
  <li><a href="#getting-python" id="toc-getting-python" class="nav-link" data-scroll-target="#getting-python">Getting Python</a></li>
  <li><a href="#choosing-an-editor" id="toc-choosing-an-editor" class="nav-link" data-scroll-target="#choosing-an-editor">Choosing an editor</a></li>
  </ul></li>
  <li><a href="#the-basics" id="toc-the-basics" class="nav-link" data-scroll-target="#the-basics">The basics</a>
  <ul class="collapse">
  <li><a href="#assigning-a-variable." id="toc-assigning-a-variable." class="nav-link" data-scroll-target="#assigning-a-variable.">Assigning a variable.</a></li>
  <li><a href="#conditionals" id="toc-conditionals" class="nav-link" data-scroll-target="#conditionals">Conditionals</a></li>
  <li><a href="#basic-plotting" id="toc-basic-plotting" class="nav-link" data-scroll-target="#basic-plotting">Basic plotting</a></li>
  </ul></li>
  <li><a href="#matricies" id="toc-matricies" class="nav-link" data-scroll-target="#matricies">Matricies</a>
  <ul class="collapse">
  <li><a href="#subsetting" id="toc-subsetting" class="nav-link" data-scroll-target="#subsetting">Subsetting</a></li>
  </ul></li>
  <li><a href="#dictionaries" id="toc-dictionaries" class="nav-link" data-scroll-target="#dictionaries">Dictionaries</a></li>
  <li><a href="#reading-and-writing-files" id="toc-reading-and-writing-files" class="nav-link" data-scroll-target="#reading-and-writing-files">Reading and writing files</a></li>
  <li><a href="#flow-control-basics" id="toc-flow-control-basics" class="nav-link" data-scroll-target="#flow-control-basics">Flow control basics</a></li>
  <li><a href="#functions" id="toc-functions" class="nav-link" data-scroll-target="#functions">Functions</a></li>
  <li><a href="#using-built-in-functions" id="toc-using-built-in-functions" class="nav-link" data-scroll-target="#using-built-in-functions">Using built-in functions</a></li>
  <li><a href="#apply" id="toc-apply" class="nav-link" data-scroll-target="#apply">Apply</a></li>
  <li><a href="#getting-highly-varable-genes." id="toc-getting-highly-varable-genes." class="nav-link" data-scroll-target="#getting-highly-varable-genes.">Getting highly varable genes.</a></li>
  <li><a href="#standardising-data" id="toc-standardising-data" class="nav-link" data-scroll-target="#standardising-data">Standardising data</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">Clustering</a>
  <ul class="collapse">
  <li><a href="#heatmaps" id="toc-heatmaps" class="nav-link" data-scroll-target="#heatmaps">Heatmaps</a></li>
  </ul></li>
  <li><a href="#linear-models" id="toc-linear-models" class="nav-link" data-scroll-target="#linear-models">Linear models</a></li>
  <li><a href="#kmeans" id="toc-kmeans" class="nav-link" data-scroll-target="#kmeans">Kmeans</a></li>
  <li><a href="#the-final-project" id="toc-the-final-project" class="nav-link" data-scroll-target="#the-final-project">The Final Project</a>
  <ul class="collapse">
  <li><a href="#optimisation-problems" id="toc-optimisation-problems" class="nav-link" data-scroll-target="#optimisation-problems">Optimisation problems</a></li>
  </ul></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>





<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>The purpose of this course is to teach you the basics of the Python language and give you the confidence to tackle larger projects using the language. <em>Importantly, we want to get you thinking like a programmer</em>. This doesn’t mean that by the end of the course you will know Python fully, but you will know <em>enough</em> so you can go online and look for the help you need to complete most tasks.</p>
<section id="practice-makes-perfect" class="level2">
<h2 class="anchored" data-anchor-id="practice-makes-perfect">Practice makes perfect</h2>
<p>Programming is like any skill, the more you practice the better you get. <strong><em>It’s really important that you keep using what you have learned after the course is completed</em></strong> otherwise there is a good chance you will forget everything and you’ll be back to square one.</p>
</section>
<section id="why-use-python" class="level2">
<h2 class="anchored" data-anchor-id="why-use-python">Why use Python?</h2>
<p>Python is a high-level <strong>general</strong> programming language (unlike R which has a focus on mathematics and statistics) so Python can be used for a wide variety of applications. When it comes to machine learning, Python seems to have the edge and is preferred by the data scientist community for it;s AI/ML capabilitioes. In bioinformatics we see Python making real in-roads as more packages become available. One of the most popular ones being <a href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> and the <a href="https://scverse.org">scverse</a> which handle single-cell analysis Python is free and available for all operating systems.</p>
</section>
<section id="which-other-languages-do-bioinformaticians-use" class="level2">
<h2 class="anchored" data-anchor-id="which-other-languages-do-bioinformaticians-use">Which other languages do bioinformaticians use?</h2>
<p><a href="https://cran.r-project.org">R</a> of course which has a long rich history in boinformatics, and most will have heard of <a href="https://www.bioconductor.org">Bioconductor</a> which is a collection of packages to anaylse biological and genomics data. For very computationally intensive tasks (e.g sequence alignment), languages such as C/C++/Rust are more commonly used, but these are far more difficult to learn.</p>
</section>
<section id="a-small-warning" class="level2">
<h2 class="anchored" data-anchor-id="a-small-warning">A small warning</h2>
<p>If you have knowledge of R, this course may be a bit frustrating as some things in Python take more effort to do. R is really geared up as a language for maths and stats, so many useful functions are build into the base language. Python is a <strong>general</strong> programming language, so functions that are native to R will need to be located in Python packages.</p>
</section>
<section id="how-will-this-course-work" class="level2">
<h2 class="anchored" data-anchor-id="how-will-this-course-work">How will this course work?</h2>
<p>We’re going to take a different approach to this course. You will be taught the basics of the Python language while doing small exercises along the way. However, we will finish by you undertaking a project which will push you quite hard. The aim is that by tacking a more difficult problem will consolidate what you have learnt, and learn more by having to look up solutions to the problems you will likely face.</p>
</section>
<section id="getting-python" class="level2">
<h2 class="anchored" data-anchor-id="getting-python">Getting Python</h2>
<p>Python is slightly different in that we make “environments” on our computers and work within these environment which makes them self-contained. You need to install Miniconda, so go <a href="https://www.anaconda.com/docs/getting-started/miniconda/main">here</a> and follow the instructions to install Miniconda on your system.</p>
<p>When you have installed Miniconda, we need to make an environment and install Python and some other packages that we need to get going. On your computer, open a terminal and do:</p>
<pre><code>conda create -n workshop.env python=3.10</code></pre>
<p>And when this is done, activate the enviroment by doing:</p>
<pre><code>conda activate workshop.env</code></pre>
<p>You are now inside the <code>workshop.env</code> environment. Anything you install/do here, will stay here.</p>
<p>Lets install some packages that we’ll need:</p>
<pre><code>conda install -c conda-forge jupyterlab notebook nbclient ipykernel pandas matplotlib numpy</code></pre>
<p>We also need to register this environment as a Jupyter kernel we will use to run commands:</p>
<pre><code>python -m ipykernel install --user --name quarto-env --display-name "Python (quarto-env)"</code></pre>
</section>
<section id="choosing-an-editor" class="level2">
<h2 class="anchored" data-anchor-id="choosing-an-editor">Choosing an editor</h2>
<p>Ask a hundred different programmer which editor is best, and you will get a hundred different answers. These are the two which people tend to recommend the most when asked:</p>
<ol type="1">
<li><p><a href="https://code.visualstudio.com">VScode</a>. This is a general purpose editor which is really popular. It will handle pretty much all languages, and plugins will make it easier to write and run code. This is the one I use, and the one we#ll use fo this course.</p></li>
<li><p><a href="https://www.jetbrains.com/pycharm/">PyCharm</a>. This is a Python specific editor which has a rediculous number of features. You can play with this in your own time!</p></li>
</ol>
</section>
</section>
<section id="the-basics" class="level1">
<h1>The basics</h1>
<p>Lets start gently and go through a few simple things to warm up.</p>
<section id="assigning-a-variable." class="level2">
<h2 class="anchored" data-anchor-id="assigning-a-variable.">Assigning a variable.</h2>
<p>Into your script copy/type the following line:</p>
<div id="7551c344" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> [<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>, <span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will make a <strong>vector</strong> of values from 1 to 10, and put them into a variable called <code>x</code>.</p>
<p>Execute the code by hitting the “Run” button at the top-right of the script window. You will see this line appear in the R console below.</p>
<p>To view the contents of the object you have just created, just type <code>x</code> in the <strong>console</strong> and hit return:</p>
<div id="e5cc4262" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="2">
<pre><code>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]</code></pre>
</div>
</div>
<p>The contents of x are now printed out.</p>
<p><strong>Now is a good time to learn about commenting and documenting code</strong>. This is free text you put into your scripts that tell the reader whats going on. It also helps remind your future self of what you did or your thought process. Comments are put in using <code>#</code>, so for example:</p>
<div id="34dd1472" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> [<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>, <span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">10</span>] <span class="co"># This is a comment</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Anything after a <code>#</code> will be ignored by Python. You can run the code again to check.</p>
<p>Rather than typing in the value 1 to 10, there is a much simpler way to create the same vector using <code>:</code></p>
<div id="e38422c5" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">11</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]</code></pre>
</div>
</div>
<p>Much better! It’s also bidirectional, so to go backwards it’s:</p>
<div id="a43ead25" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">5</span>, <span class="op">-</span><span class="dv">6</span>, <span class="op">-</span><span class="dv">1</span>))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>[5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]</code></pre>
</div>
</div>
<p>Issue the command <code>help(range)</code>. In python you can get a manual for any function using the <code>help()</code> command. Look at the help page and generate a vector of numbers from 1 to 100 in steps of 10:</p>
<div id="191905c7" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>a <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">0</span>, <span class="dv">101</span>,<span class="dv">10</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>a</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>[0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]</code></pre>
</div>
</div>
<p>Lets try something a bit more difficult. Generate a vector called <code>b</code> ranging from 3 to 987 where the length of the vector is 53 entries long. Done? Check the length of the vector you have just made.</p>
<p>In Python, there is a greater reliance on packages for working with numbers, and the package which we will use the most is <code>numpy</code>. To load a package we do:</p>
<div id="e6bd7b5e" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The way we use functions from a package is by doing <code>numpy.X</code> where <code>.X</code> is replaced by the name of the function in the <code>numpy</code> package you need. You can already see that if you need to use <code>numpy</code> functions a lot, then having to type <code>numpy</code> all the time is going to be a drag. So what we can do here is shorted it by doing:</p>
<div id="d048df92" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So now, if you want to use a <code>numpy</code> function we just use <code>np.X</code></p>
<div id="b9b49257" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>b <span class="op">=</span> np.arange(<span class="dv">3</span>, <span class="dv">938</span>, <span class="dv">53</span>)  <span class="co"># 938 is exclusive, so it matches R's inclusive 937</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>length <span class="op">=</span> <span class="bu">len</span>(b)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>18</code></pre>
</div>
</div>
<p>We can also make a new vector <code>d</code> using a vector <code>c</code>:</p>
<div id="0d51d669" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>c <span class="op">=</span> np.arange(<span class="dv">1</span>, <span class="dv">51</span>)  <span class="co"># 1 to 50 inclusive</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>d <span class="op">=</span> <span class="dv">1</span> <span class="op">/</span> c</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1.         0.5        0.33333333 0.25       0.2        0.16666667
 0.14285714 0.125      0.11111111 0.1        0.09090909 0.08333333
 0.07692308 0.07142857 0.06666667 0.0625     0.05882353 0.05555556
 0.05263158 0.05       0.04761905 0.04545455 0.04347826 0.04166667
 0.04       0.03846154 0.03703704 0.03571429 0.03448276 0.03333333
 0.03225806 0.03125    0.03030303 0.02941176 0.02857143 0.02777778
 0.02702703 0.02631579 0.02564103 0.025      0.02439024 0.02380952
 0.02325581 0.02272727 0.02222222 0.02173913 0.0212766  0.02083333
 0.02040816 0.02      ]</code></pre>
</div>
</div>
<p>And do maths on them, for example calculate the mean of <code>d</code>:</p>
<div id="33c62b65" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mean_d <span class="op">=</span> np.mean(d)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mean_d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.08998410676658848</code></pre>
</div>
</div>
<div id="058b6976" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>std_d <span class="op">=</span> np.std(d)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(std_d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.15622264625159116</code></pre>
</div>
</div>
</section>
<section id="conditionals" class="level2">
<h2 class="anchored" data-anchor-id="conditionals">Conditionals</h2>
<p>This is important stuff. If we want to ask whether something is equal to something else, we need to use the <code>==</code> operator, NOT <code>=</code>. Try this:</p>
<div id="e30d53a1" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> np.arange(<span class="dv">1</span>,<span class="dv">11</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>x <span class="op">==</span> <span class="dv">5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>array([False, False, False, False,  True, False, False, False, False,
       False])</code></pre>
</div>
</div>
<p>We can also do some other simple but important things:</p>
<div id="f4a71860" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.where(x <span class="op">&lt;</span> <span class="dv">5</span>))</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.where(x <span class="op">&lt;=</span> <span class="dv">5</span>))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.where(x <span class="op">&gt;=</span> <span class="dv">5</span>))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.where(x <span class="op">&gt;</span> <span class="dv">5</span>))</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.where(x <span class="op">!=</span> <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(array([0, 1, 2, 3]),)
(array([0, 1, 2, 3, 4]),)
(array([4, 5, 6, 7, 8, 9]),)
(array([5, 6, 7, 8, 9]),)
(array([0, 1, 2, 3, 5, 6, 7, 8, 9]),)</code></pre>
</div>
</div>
<p>Lets make another vector <code>y</code>:</p>
<div id="b5ba0fcb" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> np.arange(<span class="dv">7</span>,<span class="dv">16</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>common <span class="op">=</span> np.intersect1d(x, y)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(common)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[ 7  8  9 10]</code></pre>
</div>
</div>
<p>What do you think this does?</p>
<div id="2354bda9" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> x[<span class="op">~</span>np.isin(x, y)]</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1 2 3 4 5 6]</code></pre>
</div>
</div>
</section>
<section id="basic-plotting" class="level2">
<h2 class="anchored" data-anchor-id="basic-plotting">Basic plotting</h2>
<p>In order to plot, we need another module called <code>matplotlib</code>, specifically, a submodule called <code>pyplot</code> which can be thought of as base plotting in R (if you use R). Lets load it:</p>
<div id="79a0619d" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To plot we do:</p>
<div id="ce579722" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>plt.plot(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-19-output-1.png" width="571" height="411" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Do some googling and see how you can add a title, label the x/y axis, make the line points instead, and colour them red.</p>
<p>Here is how I would do it:</p>
<div id="f13dd797" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>plt.plot(c,d, marker<span class="op">=</span><span class="st">'o'</span>,color<span class="op">=</span><span class="st">"red"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"c"</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Index"</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"d"</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-20-output-1.png" width="589" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can make this plot a little fancier. Those who work in R will be familiar with <code>ggplot2</code> package, and there is a direct port to that in Python called <code>plotnine</code>. We use this in combaination with <code>pandas</code> which is the module that creates and handles data.frames (more on this later):</p>
<div id="b00ebaf0" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> plotnine <span class="im">import</span> ggplot, aes, geom_line, geom_point</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({<span class="st">'col1'</span>: c, <span class="st">'col2'</span>: d})</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> (</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    ggplot(df, aes(x<span class="op">=</span><span class="st">'col1'</span>, y<span class="op">=</span><span class="st">'col2'</span>)) <span class="op">+</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>    geom_line(color<span class="op">=</span><span class="st">'red'</span>) <span class="op">+</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>    geom_point(color<span class="op">=</span><span class="st">'green'</span>)</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>p.draw()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-21-output-1.png" width="678" height="486" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>If you want to do it the “pure” Python way (and we probably should!) we can to the following using <code>seaborn</code> plus <code>matplotlib</code>:</p>
<div id="6a679569" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({<span class="st">'col1'</span>: c, <span class="st">'col2'</span>: d})</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">5</span>))</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>sns.lineplot(data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'col1'</span>, y<span class="op">=</span><span class="st">'col2'</span>, color<span class="op">=</span><span class="st">'red'</span>)   <span class="co"># red line</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data<span class="op">=</span>df, x<span class="op">=</span><span class="st">'col1'</span>, y<span class="op">=</span><span class="st">'col2'</span>, color<span class="op">=</span><span class="st">'green'</span>)  <span class="co"># green points</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"1/c vs c"</span>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"col1"</span>)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"col2"</span>)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-22-output-1.png" width="663" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Lets have a look at the manual for pandas. Go <a href="https://pandas.pydata.org/docs/index.html">here</a>. This modules has extensive documentation, and at the top you can clokc on “API” which will list out all the methods which can be used on a <code>pandas</code> DataFrame.</p>
</section>
</section>
<section id="matricies" class="level1">
<h1>Matricies</h1>
<p>Matrices are the most common data format bioinformaticians work with. Microarray/RNAseq/single-cell data are all kept in matricies where gene are in the rows and samples down in the columns. Lets make a simple matrix of zeros using <code>numpy</code>:</p>
<div id="a56f4f89" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> np.zeros((<span class="dv">10</span>, <span class="dv">5</span>))  <span class="co"># 10 rows, 5 columns</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]</code></pre>
</div>
</div>
<p>This will create a matrix filled with zeros. To transpose (flip) the matrix we use t() (this will be important later!)</p>
<div id="4399537a" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>tposed_m <span class="op">=</span> m.T</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(tposed_m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]</code></pre>
</div>
</div>
<p>We usually need to name the rows and columns (genes/samples), so for that we need to switch to using a pandas dataframe because np matricies do not take labels:</p>
<div id="dd71f77a" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>    m,</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    index<span class="op">=</span>[<span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"D"</span>,<span class="st">"E"</span>,<span class="st">"F"</span>,<span class="st">"G"</span>,<span class="st">"H"</span>,<span class="st">"I"</span>,<span class="st">"J"</span>],</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>[<span class="st">"cat"</span>, <span class="st">"dog"</span>, <span class="st">"pig"</span>, <span class="st">"cow"</span>, <span class="st">"chicken"</span>]</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   cat  dog  pig  cow  chicken
A  0.0  0.0  0.0  0.0      0.0
B  0.0  0.0  0.0  0.0      0.0
C  0.0  0.0  0.0  0.0      0.0
D  0.0  0.0  0.0  0.0      0.0
E  0.0  0.0  0.0  0.0      0.0
F  0.0  0.0  0.0  0.0      0.0
G  0.0  0.0  0.0  0.0      0.0
H  0.0  0.0  0.0  0.0      0.0
I  0.0  0.0  0.0  0.0      0.0
J  0.0  0.0  0.0  0.0      0.0</code></pre>
</div>
</div>
<section id="subsetting" class="level2">
<h2 class="anchored" data-anchor-id="subsetting">Subsetting</h2>
<p>Lets make a matrix (and a vector) containing integer values so we can take a look at how subsetting work in R:</p>
<div id="25a1e05f" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#m = np.arange(1, 51).reshape((5, 10), order='F').T  # 10x5 after transpose</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>v <span class="op">=</span> np.arange(<span class="dv">1</span>, <span class="dv">11</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> np.arange(<span class="dv">1</span>, <span class="dv">51</span>).reshape((<span class="dv">10</span>, <span class="dv">5</span>))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    m,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    index<span class="op">=</span>[<span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"D"</span>,<span class="st">"E"</span>,<span class="st">"F"</span>,<span class="st">"G"</span>,<span class="st">"H"</span>,<span class="st">"I"</span>,<span class="st">"J"</span>],</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>[<span class="st">"cat"</span>, <span class="st">"dog"</span>, <span class="st">"pig"</span>, <span class="st">"cow"</span>, <span class="st">"chicken"</span>]</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   cat  dog  pig  cow  chicken
A    1    2    3    4        5
B    6    7    8    9       10
C   11   12   13   14       15
D   16   17   18   19       20
E   21   22   23   24       25
F   26   27   28   29       30
G   31   32   33   34       35
H   36   37   38   39       40
I   41   42   43   44       45
J   46   47   48   49       50</code></pre>
</div>
</div>
<p>We can access individual elements using square brackets <code>[]</code>. So to get the 6th, 1st and 5th elements of <code>v</code> we need:</p>
<div id="c15f8fac" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(v[[<span class="dv">6</span>,<span class="dv">0</span>,<span class="dv">4</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[7 1 5]</code></pre>
</div>
</div>
<p>Why? Because <strong>Python counts from 0.</strong> If you have been using R until now, you will know it is one based. To get the first element of a vector in R you would d <code>v[1]</code>, but this isn’t so in Python. In fact, pretty much all languages are 0 based. Wit this in mind, lets look ar the first row of the matrix <code>m</code>.</p>
<div id="dc5f6171" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.iloc[<span class="dv">0</span>,:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cat        1
dog        2
pig        3
cow        4
chicken    5
Name: A, dtype: int64</code></pre>
</div>
</div>
<p>and the 3rd column:</p>
<div id="2e2f5e9e" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.iloc[:,<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A     3
B     8
C    13
D    18
E    23
F    28
G    33
H    38
I    43
J    48
Name: pig, dtype: int64</code></pre>
</div>
</div>
<div id="502fe0b9" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.iloc[:,[<span class="dv">1</span>,<span class="dv">4</span>]]) <span class="co"># the 2nd and 5th column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   dog  chicken
A    2        5
B    7       10
C   12       15
D   17       20
E   22       25
F   27       30
G   32       35
H   37       40
I   42       45
J   47       50</code></pre>
</div>
</div>
<div id="f619ac5c" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.iloc[:,[<span class="dv">1</span>,<span class="dv">4</span>]]) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   dog  chicken
A    2        5
B    7       10
C   12       15
D   17       20
E   22       25
F   27       30
G   32       35
H   37       40
I   42       45
J   47       50</code></pre>
</div>
</div>
<div id="955ff294" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.iloc[<span class="dv">2</span>:<span class="dv">7</span>, <span class="dv">3</span>]) <span class="co"># the 3rd to 7th row of the 4th column. Remember 7 is exclusive</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>C    14
D    19
E    24
F    29
G    34
Name: cow, dtype: int64</code></pre>
</div>
</div>
<div id="b792b6ab" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.loc[<span class="st">"B"</span>,:]) <span class="co"># gets the row labelled B</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cat         6
dog         7
pig         8
cow         9
chicken    10
Name: B, dtype: int64</code></pre>
</div>
</div>
<div id="cbd5d33a" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.loc[<span class="st">"B"</span>,<span class="st">"cow"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>9</code></pre>
</div>
</div>
<div id="b70b20f2" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(df.loc[[<span class="st">"F"</span>, <span class="st">"J"</span>],[<span class="st">"chicken"</span>, <span class="st">"cat"</span>, <span class="st">"pig"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   chicken  cat  pig
F       30   26   28
J       50   46   48</code></pre>
</div>
</div>
<p>We often need to collect vectors and assemble them into a matrix. This can be done using the rbind (row) and cbind (column) functions:</p>
<div id="3010ba10" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>v1 <span class="op">=</span> np.arange(<span class="dv">1</span>, <span class="dv">11</span>)         <span class="co"># [1, 2, ..., 10]</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>v2 <span class="op">=</span> np.arange(<span class="dv">101</span>, <span class="dv">111</span>)      <span class="co"># [101, 102, ..., 110]</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>rbound_mat <span class="op">=</span> np.vstack([v1, v2])</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(rbound_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[  1   2   3   4   5   6   7   8   9  10]
 [101 102 103 104 105 106 107 108 109 110]]</code></pre>
</div>
</div>
</section>
</section>
<section id="dictionaries" class="level1">
<h1>Dictionaries</h1>
<p>So far we have talked about vectors and matrices separately, but we often we want to collect these things and put them into one object under a single variable as a collection (for example expression data and gene annotation). To do this we use something called a <strong>dictionary</strong>.</p>
<div id="6fd19521" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>alpha <span class="op">=</span> [<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>, <span class="st">"F"</span>, <span class="st">"G"</span>, <span class="st">"H"</span>]</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>mat <span class="op">=</span> np.random.randn(<span class="dv">8</span>, <span class="dv">5</span>)  <span class="co"># 8 rows, 5 columns of random normal values</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>listex1 <span class="op">=</span> {</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"char"</span>: alpha,</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"nums"</span>: mat</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can see that each item is given a name <code>char</code> <code>nums</code> before it is put into the list. Each element can now be accessed via <code>[]</code>:</p>
<div id="703abf48" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>listex1[<span class="st">"char"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="37">
<pre><code>['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']</code></pre>
</div>
</div>
<p>So to access the 3rd element of the vector:</p>
<div id="b28a25e1" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>listex1[<span class="st">"char"</span>][<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>'C'</code></pre>
</div>
</div>
<p>And the first row of the dataframe:</p>
<div id="32a7f83d" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>listex1[<span class="st">"nums"</span>][<span class="dv">0</span>,:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="39">
<pre><code>array([ 1.7967597 ,  0.09344559,  0.8266841 , -0.0961537 , -1.14112448])</code></pre>
</div>
</div>
<div id="1d7a9299" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.datasets <span class="im">import</span> load_iris</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>iris_data <span class="op">=</span> load_iris(as_frame<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>iris <span class="op">=</span> iris_data.frame  <span class="co"># pandas DataFrame</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample 10 random rows</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>iris.head(<span class="dv">5</span>)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="co">#type(iris)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="40">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">petal length (cm)</th>
<th data-quarto-table-cell-role="th">petal width (cm)</th>
<th data-quarto-table-cell-role="th">target</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.1</td>
<td>3.5</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>4.9</td>
<td>3.0</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>1.3</td>
<td>0.2</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4.6</td>
<td>3.1</td>
<td>1.5</td>
<td>0.2</td>
<td>0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5.0</td>
<td>3.6</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="539cbc48" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>plt.scatter(iris[<span class="st">"sepal length (cm)"</span>], iris[<span class="st">"sepal width (cm)"</span>])</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Sepal Length"</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Sepal Width"</span>)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Sepal Length vs Sepal Width"</span>)</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-42-output-1.png" width="589" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="365d787f" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>iris[<span class="st">"species"</span>] <span class="op">=</span> iris_data.target_names[iris_data.target]</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the plot</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(data<span class="op">=</span>iris, x<span class="op">=</span><span class="st">"sepal length (cm)"</span>, y<span class="op">=</span><span class="st">"sepal width (cm)"</span>, hue<span class="op">=</span><span class="st">"species"</span>)</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Sepal Length vs Sepal Width"</span>)</span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-43-output-1.png" width="589" height="449" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can also break up a data frame up on attributes and store the contents in a list (which we have already discussed). For example by species:</p>
<div id="5e20e75e" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>iris_sp <span class="op">=</span> {species: data <span class="cf">for</span> species, data <span class="kw">in</span> iris.groupby(<span class="st">"species"</span>)}</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the keys (species names) of the dictionary</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="bu">list</span>(iris_sp.keys())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="43">
<pre><code>['setosa', 'versicolor', 'virginica']</code></pre>
</div>
</div>
<div id="32e0e2b6" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>iris_sp[<span class="st">'setosa'</span>].iloc[<span class="dv">0</span>:<span class="dv">3</span>, :<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="44">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">petal length (cm)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.1</td>
<td>3.5</td>
<td>1.4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>4.9</td>
<td>3.0</td>
<td>1.4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>1.3</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="d9ea7e99" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>iris_sp[<span class="st">'versicolor'</span>].iloc[<span class="dv">0</span>:<span class="dv">3</span>, :<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="45">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">petal length (cm)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">50</td>
<td>7.0</td>
<td>3.2</td>
<td>4.7</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">51</td>
<td>6.4</td>
<td>3.2</td>
<td>4.5</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">52</td>
<td>6.9</td>
<td>3.1</td>
<td>4.9</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Let do some basic filtering and manipulation of this data. First, filter just for those which are setosa:</p>
<div id="c30e872c" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>setosa_data <span class="op">=</span> iris[iris[<span class="st">"species"</span>] <span class="op">==</span> <span class="st">"setosa"</span>]</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>setosa_data.iloc[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="46">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">petal length (cm)</th>
<th data-quarto-table-cell-role="th">petal width (cm)</th>
<th data-quarto-table-cell-role="th">target</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.1</td>
<td>3.5</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>4.9</td>
<td>3.0</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>1.3</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4.6</td>
<td>3.1</td>
<td>1.5</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5.0</td>
<td>3.6</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Select 3 specific columsn of the data</p>
<div id="598e3914" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>iris_subset <span class="op">=</span> iris[[<span class="st">"sepal length (cm)"</span>, <span class="st">"sepal width (cm)"</span>, <span class="st">"species"</span>]]</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>iris_subset.iloc[:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="47">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.1</td>
<td>3.5</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>4.9</td>
<td>3.0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4.6</td>
<td>3.1</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5.0</td>
<td>3.6</td>
<td>setosa</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Sort on a field:</p>
<div id="f8962766" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>iris_sorted <span class="op">=</span> iris.sort_values(<span class="st">"sepal length (cm)"</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>iris_sorted.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="48">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sepal length (cm)</th>
<th data-quarto-table-cell-role="th">sepal width (cm)</th>
<th data-quarto-table-cell-role="th">petal length (cm)</th>
<th data-quarto-table-cell-role="th">petal width (cm)</th>
<th data-quarto-table-cell-role="th">target</th>
<th data-quarto-table-cell-role="th">species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">13</td>
<td>4.3</td>
<td>3.0</td>
<td>1.1</td>
<td>0.1</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">8</td>
<td>4.4</td>
<td>2.9</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">42</td>
<td>4.4</td>
<td>3.2</td>
<td>1.3</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">38</td>
<td>4.4</td>
<td>3.0</td>
<td>1.3</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">41</td>
<td>4.5</td>
<td>2.3</td>
<td>1.3</td>
<td>0.3</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">22</td>
<td>4.6</td>
<td>3.6</td>
<td>1.0</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>4.6</td>
<td>3.1</td>
<td>1.5</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6</td>
<td>4.6</td>
<td>3.4</td>
<td>1.4</td>
<td>0.3</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47</td>
<td>4.6</td>
<td>3.2</td>
<td>1.4</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>1.3</td>
<td>0.2</td>
<td>0</td>
<td>setosa</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="reading-and-writing-files" class="level1">
<h1>Reading and writing files</h1>
<p>You have to get the data into R first before you can analyse it (this helps a lot). R has many useful functions to do this, so now we can take our first look at some expression data. Download this <a href="https://github.com/sccbioinformatics/Python_programming_1/blob/main/Mouse_HSPC_reduced.txt">file</a> and save it to your current working directory.</p>
<p><strong><em>Exercise</em></strong>: Open the file in Excel or something to see how it looks. As we want a <strong>named</strong> matrix we know that we want a <code>pandas</code> dataframe. So call <code>help(pd.read_csv)</code> in the terminal and try to work out how to import this file.</p>
<p>This is how I would do it:</p>
<div id="d064bd1c" class="cell" data-execution_count="49">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>hspc_data <span class="op">=</span> pd.read_csv(<span class="st">"Mouse_HSPC_reduced.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, header<span class="op">=</span><span class="dv">0</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Lets look are the first few lines:</p>
<div id="8a7a1e0f" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>hspc_data.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">LTHSC.1</th>
<th data-quarto-table-cell-role="th">LTHSC.2</th>
<th data-quarto-table-cell-role="th">LTHSC.3</th>
<th data-quarto-table-cell-role="th">LTHSC.4</th>
<th data-quarto-table-cell-role="th">LTHSC.5</th>
<th data-quarto-table-cell-role="th">LTHSC.6</th>
<th data-quarto-table-cell-role="th">LTHSC.7</th>
<th data-quarto-table-cell-role="th">LTHSC.8</th>
<th data-quarto-table-cell-role="th">LTHSC.9</th>
<th data-quarto-table-cell-role="th">LTHSC.10</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">GMP.41</th>
<th data-quarto-table-cell-role="th">GMP.42</th>
<th data-quarto-table-cell-role="th">GMP.43</th>
<th data-quarto-table-cell-role="th">GMP.44</th>
<th data-quarto-table-cell-role="th">GMP.45</th>
<th data-quarto-table-cell-role="th">GMP.46</th>
<th data-quarto-table-cell-role="th">GMP.47</th>
<th data-quarto-table-cell-role="th">GMP.48</th>
<th data-quarto-table-cell-role="th">GMP.49</th>
<th data-quarto-table-cell-role="th">GMP.50</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Kdm3a</td>
<td>0.000000</td>
<td>7.326561</td>
<td>1.141277</td>
<td>0.000000</td>
<td>0.000000</td>
<td>2.969495</td>
<td>0.000000</td>
<td>3.936049</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.000000</td>
<td>4.853605</td>
<td>1.921055</td>
<td>6.008491</td>
<td>6.501955</td>
<td>4.240413</td>
<td>1.074001</td>
<td>1.306210</td>
<td>0.000000</td>
<td>1.277098</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Coro2b</td>
<td>1.312809</td>
<td>5.699220</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>6.115329</td>
<td>0.000000</td>
<td>0.000000</td>
<td>7.606749</td>
<td>0.000000</td>
<td>...</td>
<td>1.022269</td>
<td>0.331643</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>8.062927</td>
<td>1.074001</td>
<td>0.796145</td>
<td>6.726593</td>
<td>0.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Phf6</td>
<td>0.000000</td>
<td>1.685035</td>
<td>6.254748</td>
<td>2.369840</td>
<td>2.158438</td>
<td>7.823217</td>
<td>7.735619</td>
<td>1.111009</td>
<td>6.217745</td>
<td>3.919389</td>
<td>...</td>
<td>1.022269</td>
<td>8.521508</td>
<td>4.542010</td>
<td>3.365015</td>
<td>1.534551</td>
<td>8.193208</td>
<td>3.298967</td>
<td>6.573894</td>
<td>5.297550</td>
<td>7.745012</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Usp14</td>
<td>1.312809</td>
<td>3.301869</td>
<td>1.770417</td>
<td>3.223063</td>
<td>6.106443</td>
<td>8.544492</td>
<td>9.792677</td>
<td>8.979602</td>
<td>5.821215</td>
<td>1.835594</td>
<td>...</td>
<td>5.658020</td>
<td>7.418338</td>
<td>5.826294</td>
<td>7.360951</td>
<td>8.439620</td>
<td>1.635074</td>
<td>7.352915</td>
<td>8.031799</td>
<td>8.969543</td>
<td>9.187885</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Tmem167b</td>
<td>4.233503</td>
<td>0.000000</td>
<td>0.680667</td>
<td>1.624999</td>
<td>0.000000</td>
<td>2.969495</td>
<td>7.900999</td>
<td>0.471609</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.000000</td>
<td>3.171820</td>
<td>1.921055</td>
<td>7.684185</td>
<td>4.419037</td>
<td>1.037746</td>
<td>1.682810</td>
<td>0.796145</td>
<td>0.000000</td>
<td>6.525201</td>
</tr>
</tbody>
</table>

<p>5 rows × 150 columns</p>
</div>
</div>
</div>
<p>Lets have a look at the dimensions of the data. In the <code>pandas</code> module we have the <code>shape</code> function to get this:</p>
<div id="51c34f1b" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>hspc_data.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="51">
<pre><code>(4170, 150)</code></pre>
</div>
</div>
<p>We can access the individal components of this using:</p>
<div id="c3b118e6" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>hspc_data.shape[<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="52">
<pre><code>4170</code></pre>
</div>
</div>
<div id="84bc8353" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>hspc_data.shape[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="53">
<pre><code>150</code></pre>
</div>
</div>
<p>Lets take a look at the column names and see what they look like:</p>
<div id="8fa8d6b6" class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="bu">list</span>(hspc_data.columns)[<span class="dv">1</span>:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="54">
<pre><code>['LTHSC.2',
 'LTHSC.3',
 'LTHSC.4',
 'LTHSC.5',
 'LTHSC.6',
 'LTHSC.7',
 'LTHSC.8',
 'LTHSC.9',
 'LTHSC.10']</code></pre>
</div>
</div>
<p>It’s quite common that we need to break a table up. In this case, we want to separate the different populations into dataframes called <code>lthsc</code>,<code>mep</code>, and <code>gmp</code>.</p>
<p>To do this we need to get to the string atributes of the column headers, and to do this we need to use <code>.str</code> and the functions contained within. an example of this:</p>
<div id="bbc37324" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>mep_data <span class="op">=</span> hspc_data.loc[:, hspc_data.columns.<span class="bu">str</span>.contains(<span class="st">"MEP"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let break this down inside out. <code>hspc_data.columns.str.contains("MEP")</code> accessing the string functions using <code>.str</code>, and then uses the <code>.contains</code> function to ask which of the <code>.columns</code> contains “MEP”. We then pass the results of this to <code>hspc_data.loc</code> to pull out the corresponsing columns.</p>
<p>Lets run the inner part by itself:</p>
<div id="29490202" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>hspc_data.columns.<span class="bu">str</span>.contains(<span class="st">"MEP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="56">
<pre><code>array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False])</code></pre>
</div>
</div>
<p>We can also get the same results by using <code>.startswith</code> instead:</p>
<div id="d3ee45d0" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>hspc_data.columns.<span class="bu">str</span>.startswith(<span class="st">"MEP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="57">
<pre><code>array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False])</code></pre>
</div>
</div>
<p><strong><em>Exercise</em></strong>: Isolate the data for the LTHSC and GMP data into objects called <code>lthsc_data</code> and <code>gmp_data</code>.</p>
<div id="e595815a" class="cell" data-execution_count="58">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>lthsc_data <span class="op">=</span> hspc_data.loc[:, hspc_data.columns.<span class="bu">str</span>.contains(<span class="st">"LTHSC"</span>)]</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>gmp_data <span class="op">=</span> hspc_data.loc[:, hspc_data.columns.<span class="bu">str</span>.contains(<span class="st">"GMP"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let have a look at another important task, writing out a file.</p>
<p><strong><em>Exercise</em></strong>: Call <code>pd.DataFrame.to_csv</code> and have a look at the help page. Try and figure out how to write the <code>lthsc_data</code> to a file called <code>LTHSC_data.txt</code>.</p>
<div id="e2d1f42d" class="cell" data-execution_count="59">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>lthsc_data.to_csv(<span class="st">"LTHSC_data.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">True</span>, header<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong><em>Exercise</em></strong>: Write out the other two files!</p>
<div id="6a8576a8" class="cell" data-execution_count="60">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>mep_data.to_csv(<span class="st">"MEP_data.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">True</span>, header<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>gmp_data.to_csv(<span class="st">"GMP_data.txt"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">True</span>, header<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="flow-control-basics" class="level1">
<h1>Flow control basics</h1>
<p>This is where things get more interesting and we start to feel like “proper” programmers. Now that we have these datasets loaded in Python, we can use them to learn about flow control and some basic mathematical functions. We are going to do a few things the “long way” so you get the idea of how flow control works, and then we’ll look at some shortcuts.</p>
<p>Flow control is how multi-step processes (such as algorithms) are carried out. In the example below we print out the numbers 1 to 10:</p>
<div id="35d74968" class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">10</span>):</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(i)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0
1
2
3
4
5
6
7
8
9</code></pre>
</div>
</div>
<p>To translate this code, it simply says for every integer from 0 to 9, print this value to the screen.</p>
<p><strong><em>Exercises</em></strong>:</p>
<ol type="1">
<li>Using the example above, print the first 10 lines of ‘lthcs_data’ in a for loop.</li>
</ol>
<div id="63203652" class="cell" data-execution_count="62">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">10</span>):</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(hspc_data.iloc[i,:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ol start="2" type="1">
<li>Print every 2nd line of <code>mep_data</code> from lines 0 to 49.</li>
</ol>
<div id="71ce4a34" class="cell" data-execution_count="63">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>,<span class="dv">50</span>,<span class="dv">2</span>):</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(hspc_data.iloc[i,:])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>An important point regarding for loops is that any processes/calculations occurring within the loop will stay in the loop. If data generated within a loop has to be retained, we need to create a container to “fill up” while the loop is being carried out. for example:</p>
<div id="07db6c01" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>vec <span class="op">=</span> []</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">10</span>):</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>    vec.append(i <span class="op">*</span> <span class="dv">10</span>)</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(vec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]</code></pre>
</div>
</div>
<p>The empty container <code>vec</code> is initialised outside the loop, and then populated by concatenating to it using <code>append</code> after every iteration of the loop.</p>
<p><strong><em>Exercise</em></strong>: Initialise an empty container, and for <code>gmp_data</code>, calculate the mean of each row (gene), and store the results in the container you made.</p>
<div id="e5c42139" class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>gmp_row_means <span class="op">=</span> []</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(gmp_data.shape[<span class="dv">0</span>]):</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>    gmp_row_means.append(gmp_data.iloc[i,:].mean())</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(gmp_row_means))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another method of flow control is <code>while</code>. for example:</p>
<div id="8b5504be" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>w <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> (w <span class="op">&lt;=</span> <span class="dv">5</span>):</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(w)</span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>    w <span class="op">=</span> w <span class="op">+</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0
1
2
3
4
5</code></pre>
</div>
</div>
<p>We will cover <code>if</code> and <code>else</code> a bit later.</p>
</section>
<section id="functions" class="level1">
<h1>Functions</h1>
<p>Functions are chunks of code that execute several lines of code at once to perform a task. Once you have a few lines of useful code that you want to apply repeatedly, a function is a nice way to wrap them up so it can be used quickly when needed. Lets turn the code you wrote in the previous exercise into a function where we also calculate the variance for a gene.</p>
<div id="2df64c46" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calc_mean_and_var(mat):</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>    means <span class="op">=</span> []</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>    variances <span class="op">=</span> []</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(mat.shape[<span class="dv">0</span>]):</span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>        row <span class="op">=</span> mat.iloc[i, :]</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>        means.append(np.mean(row))</span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a>        variances.append(np.var(row))</span>
<span id="cb110-9"><a href="#cb110-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-10"><a href="#cb110-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {<span class="st">"mns"</span>: means, <span class="st">"vars"</span>: variances} <span class="co"># This line gives us back what we need.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, you can see a loop is started, and the output from each loop is appended to containers <code>means</code> and <code>variances</code>. These are both put into a list which is returned at the end. By putting this code into a function we can now calculate the means and deviations of any matrix in one line of code. For example, the <code>gmp</code> data:</p>
<div id="d97b96af" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>gmp_mn_var <span class="op">=</span> calc_mean_and_var(gmp_data)</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gmp_mn_var[<span class="st">"mns"</span>])</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gmp_mn_var[<span class="st">"vars"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also use functions to demonstrate <code>if</code> and <code>else</code> statements reagarding flow control. Lets have a look at this using a silly example:</p>
<div id="1cc61af5" class="cell" data-execution_count="69">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> animal_maths(value1, value2, animal<span class="op">=</span><span class="st">"pig"</span>):</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> animal <span class="op">==</span> <span class="st">"pig"</span>:</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(value1 <span class="op">/</span> value2)</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> animal <span class="op">==</span> <span class="st">"cow"</span>:</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(value1 <span class="op">*</span> value2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>… and run it:</p>
<div id="0a751e6d" class="cell" data-execution_count="70">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"pig"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2.0</code></pre>
</div>
</div>
<div id="9b7fd63e" class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"cow"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>50</code></pre>
</div>
</div>
<p>What happens if we do this?</p>
<div id="440e20b6" class="cell" data-execution_count="72">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"dog"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What we need to do here is a little “exemption handling”. You can’t rely on the user doing the correct thing, so we need to account for any odd things that happen. In this case, the user has put an animal into the function which isn’t accounted for by the function. Let’s modify the function to let the user know they screwed up:</p>
<div id="cd92937f" class="cell" data-execution_count="73">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> animal_maths(value1, value2, animal<span class="op">=</span><span class="st">"pig"</span>):</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> animal <span class="op">==</span> <span class="st">"pig"</span>:</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(value1 <span class="op">/</span> value2)</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> animal <span class="op">==</span> <span class="st">"cow"</span>:</span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(value1 <span class="op">*</span> value2)</span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:(<span class="bu">print</span>(<span class="st">"I don't recognise this animal!"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c1ceb661" class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"dog"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>I don't recognise this animal!</code></pre>
</div>
</div>
<p>We’re using Python, so we should do this the more “Pythonic” way, and there is a really cool way of doing this using a dictionary to lookup what needs to be done:</p>
<div id="fd7d9d26" class="cell" data-execution_count="75">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> animal_maths(value1, value2, animal<span class="op">=</span><span class="st">"pig"</span>):</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>    operations <span class="op">=</span> {</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"pig"</span>: <span class="kw">lambda</span> x, y: x <span class="op">/</span> y,</span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"cow"</span>: <span class="kw">lambda</span> x, y: x <span class="op">*</span> y</span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> animal <span class="kw">in</span> operations:</span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(operations[animal](value1, value2))</span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"Unknown animal"</span>)</span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-11"><a href="#cb121-11" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"pig"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2.0</code></pre>
</div>
</div>
<p>What we do here is provide the options in a dictionary where each one is an unnamed function (lambda function), so all we need is to provide one <code>if</code> and one <code>else</code>, as opposed to <code>if</code> &gt; <code>elif</code> &gt; <code>else</code>.</p>
<div id="869ab0e8" class="cell" data-execution_count="76">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>animal_maths(<span class="dv">10</span>, <span class="dv">5</span>, <span class="st">"cow"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>50</code></pre>
</div>
</div>
</section>
<section id="using-built-in-functions" class="level1">
<h1>Using built-in functions</h1>
<p>Libraries such as <code>pandas</code> have a host of functions that can be applied to <code>pandas</code> DataFrames. Let have a look at this code which calculated row means and variances of a matrix:</p>
<div id="1260276a" class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calc_mean_and_var(mat):</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>    means <span class="op">=</span> []</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>    variances <span class="op">=</span> []</span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(mat.shape[<span class="dv">0</span>]): <span class="co"># &lt;&lt;&lt; This is slow!</span></span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>        row <span class="op">=</span> mat.iloc[i, :]</span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>        means.append(np.mean(row))</span>
<span id="cb125-8"><a href="#cb125-8" aria-hidden="true" tabindex="-1"></a>        variances.append(np.var(row))</span>
<span id="cb125-9"><a href="#cb125-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb125-10"><a href="#cb125-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {<span class="st">"mns"</span>: means, <span class="st">"vars"</span>: variances} </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function loops through each line which is computationally <em>really</em> slow as each line had to be interpreted in turn. Lets try this instead:</p>
<div id="e6829766" class="cell" data-execution_count="78">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>gmp_row_means <span class="op">=</span> gmp_data.mean(axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is so much faster.</p>
<p><strong><em>Exercise</em></strong>: Write a function called <code>calc_mean_and_var_fast</code> which uses this method to calculate row means and variances for a matrix.</p>
<div id="4f2c456f" class="cell" data-execution_count="79">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> calc_mean_and_var_fast(mat):</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {</span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"mns"</span>: mat.mean(axis<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"vars"</span>: mat.var(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a>    }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use the <code>time</code> library to measure how long it takes to run a functions. Slow one first:</p>
<div id="1cd02515" class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> calc_mean_and_var(gmp_data)  <span class="co"># Call your function</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb128-6"><a href="#cb128-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb128-7"><a href="#cb128-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Time taken: </span><span class="sc">{</span>end_time <span class="op">-</span> start_time<span class="sc">:.6f}</span><span class="ss"> seconds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time taken: 0.571316 seconds</code></pre>
</div>
</div>
<p>Now the fast version:</p>
<div id="f832f955" class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> calc_mean_and_var_fast(gmp_data)  <span class="co"># Call your function</span></span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Time taken: </span><span class="sc">{</span>end_time <span class="op">-</span> start_time<span class="sc">:.6f}</span><span class="ss"> seconds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time taken: 0.005224 seconds</code></pre>
</div>
</div>
</section>
<section id="apply" class="level1">
<h1>Apply</h1>
<p><code>apply</code> is a commonly used function in Python to speed up matrix calculation when you have a custom function that you would like to use. For eample, lets take a nonsense operation:</p>
<div id="a59139fe" class="cell" data-execution_count="82">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> example_func(v):</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>    val <span class="op">=</span> (np.mean(v) <span class="op">*</span> np.std(v)) <span class="op">/</span> np.<span class="bu">sum</span>(v)</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> val</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>ex_apply <span class="op">=</span> gmp_data.<span class="bu">apply</span>(example_func, axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll likely use this a bit later!</p>
</section>
<section id="getting-highly-varable-genes." class="level1">
<h1>Getting highly varable genes.</h1>
<p>Lets start doing some work with this data, and first we’ll define a function that isloates the <em>N</em> highest varable genes in the <code>hspc_data</code>.</p>
<div id="d535aa19" class="cell" data-execution_count="83">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_top_variable_genes(mat, top_n<span class="op">=</span><span class="dv">500</span>):</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>    variances <span class="op">=</span> mat.var(axis<span class="op">=</span><span class="dv">1</span>, ddof<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>    top_genes <span class="op">=</span> variances.nlargest(top_n).index</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> mat.loc[top_genes]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lets run it.</p>
<div id="5833c6fc" class="cell" data-execution_count="84">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>hspc_var <span class="op">=</span> get_top_variable_genes(hspc_data)</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>hspc_var.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="80">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">LTHSC.1</th>
<th data-quarto-table-cell-role="th">LTHSC.2</th>
<th data-quarto-table-cell-role="th">LTHSC.3</th>
<th data-quarto-table-cell-role="th">LTHSC.4</th>
<th data-quarto-table-cell-role="th">LTHSC.5</th>
<th data-quarto-table-cell-role="th">LTHSC.6</th>
<th data-quarto-table-cell-role="th">LTHSC.7</th>
<th data-quarto-table-cell-role="th">LTHSC.8</th>
<th data-quarto-table-cell-role="th">LTHSC.9</th>
<th data-quarto-table-cell-role="th">LTHSC.10</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">GMP.41</th>
<th data-quarto-table-cell-role="th">GMP.42</th>
<th data-quarto-table-cell-role="th">GMP.43</th>
<th data-quarto-table-cell-role="th">GMP.44</th>
<th data-quarto-table-cell-role="th">GMP.45</th>
<th data-quarto-table-cell-role="th">GMP.46</th>
<th data-quarto-table-cell-role="th">GMP.47</th>
<th data-quarto-table-cell-role="th">GMP.48</th>
<th data-quarto-table-cell-role="th">GMP.49</th>
<th data-quarto-table-cell-role="th">GMP.50</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Elane</td>
<td>4.939460</td>
<td>6.685269</td>
<td>3.893970</td>
<td>0.000000</td>
<td>3.219455</td>
<td>0.000000</td>
<td>1.954717</td>
<td>0.471609</td>
<td>2.924017</td>
<td>0.000000</td>
<td>...</td>
<td>13.174443</td>
<td>1.350880</td>
<td>13.216534</td>
<td>9.535947</td>
<td>14.670066</td>
<td>5.745529</td>
<td>12.743118</td>
<td>9.285498</td>
<td>8.952020</td>
<td>12.874189</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Ctsg</td>
<td>4.939460</td>
<td>6.374567</td>
<td>4.059607</td>
<td>4.701162</td>
<td>4.618081</td>
<td>3.491716</td>
<td>1.954717</td>
<td>1.731154</td>
<td>3.918433</td>
<td>1.835594</td>
<td>...</td>
<td>13.425062</td>
<td>3.212612</td>
<td>12.257906</td>
<td>12.745897</td>
<td>14.029701</td>
<td>8.644193</td>
<td>11.430893</td>
<td>12.536193</td>
<td>11.353980</td>
<td>12.940480</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Mpo</td>
<td>6.018427</td>
<td>8.361467</td>
<td>5.690473</td>
<td>5.673162</td>
<td>6.508938</td>
<td>6.433089</td>
<td>3.801551</td>
<td>3.306419</td>
<td>5.721786</td>
<td>2.617855</td>
<td>...</td>
<td>14.345693</td>
<td>5.035134</td>
<td>13.778882</td>
<td>13.803313</td>
<td>15.709696</td>
<td>11.056773</td>
<td>13.921738</td>
<td>13.472434</td>
<td>12.350364</td>
<td>14.891752</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Car1</td>
<td>5.072683</td>
<td>7.564652</td>
<td>2.541775</td>
<td>5.267519</td>
<td>6.552186</td>
<td>8.081555</td>
<td>3.267415</td>
<td>11.992240</td>
<td>2.732963</td>
<td>6.623940</td>
<td>...</td>
<td>3.948185</td>
<td>8.337803</td>
<td>8.328697</td>
<td>3.845319</td>
<td>3.835810</td>
<td>3.653425</td>
<td>4.597563</td>
<td>5.521170</td>
<td>5.297550</td>
<td>5.584235</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Ms4a3</td>
<td>2.446983</td>
<td>3.301869</td>
<td>1.141277</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.781653</td>
<td>0.826445</td>
<td>0.957072</td>
<td>1.191937</td>
<td>...</td>
<td>10.853943</td>
<td>0.000000</td>
<td>8.932879</td>
<td>7.511054</td>
<td>11.256612</td>
<td>2.871469</td>
<td>9.112783</td>
<td>8.701691</td>
<td>9.473927</td>
<td>10.380899</td>
</tr>
</tbody>
</table>

<p>5 rows × 150 columns</p>
</div>
</div>
</div>
</section>
<section id="standardising-data" class="level1">
<h1>Standardising data</h1>
<p>Lets stick with our expression data and cluster it. Doing so, we’ll learn more of the language, and some of the fundamental maths that runs underneath. Lets take a look at the range of the data, i.e.&nbsp;getting the lowest and highest value in the matrix of variable genes we just made.</p>
<div id="56065dfd" class="cell" data-execution_count="85">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>hspc_var.values.<span class="bu">min</span>(), hspc_var.values.<span class="bu">max</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="81">
<pre><code>(np.float64(0.0), np.float64(15.7096959783704))</code></pre>
</div>
</div>
<p>For some operations (such as making heatmaps) the data needs to be z-score normalised (scaled) first. When we scale data, each row of gene is standarised so that it’s mean = 0 and sd = 1. Specifically for a gene <code>g</code> of the i-th row:</p>
<p><span class="math display">\[Z_i=\frac{g_i-\hat{g}}{\sigma g_i}\]</span></p>
<p>Which means for every gene (<span class="math inline">\(g_i\)</span>) we subtract the mean expression of the gene (<span class="math inline">\(\hat{g)}\)</span>) to each expression value (<span class="math inline">\(g_i\)</span>) and divide by the standard deviation of the expression of the gene (<span class="math inline">\(\sigma g_i\)</span>).</p>
<p><strong><em>Exercise</em></strong>: Write a function called <code>zscore</code> which will take a single vector of values <code>v</code> and scale them. When you have done this, apply this to the hspc.var matrix to scale all rows and call it <code>hspc_zs</code>.</p>
<div id="724d614d" class="cell" data-execution_count="86">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> zscore(v):</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>    zv <span class="op">=</span> (v <span class="op">-</span> np.mean(v))<span class="op">/</span>(np.std(v,ddof<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> zv</span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>hspc_zs <span class="op">=</span> hspc_var.<span class="bu">apply</span>(zscore,<span class="dv">1</span>)</span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(hspc_zs.shape)</span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a>hspc_zs.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(500, 150)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="82">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">LTHSC.1</th>
<th data-quarto-table-cell-role="th">LTHSC.2</th>
<th data-quarto-table-cell-role="th">LTHSC.3</th>
<th data-quarto-table-cell-role="th">LTHSC.4</th>
<th data-quarto-table-cell-role="th">LTHSC.5</th>
<th data-quarto-table-cell-role="th">LTHSC.6</th>
<th data-quarto-table-cell-role="th">LTHSC.7</th>
<th data-quarto-table-cell-role="th">LTHSC.8</th>
<th data-quarto-table-cell-role="th">LTHSC.9</th>
<th data-quarto-table-cell-role="th">LTHSC.10</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">GMP.41</th>
<th data-quarto-table-cell-role="th">GMP.42</th>
<th data-quarto-table-cell-role="th">GMP.43</th>
<th data-quarto-table-cell-role="th">GMP.44</th>
<th data-quarto-table-cell-role="th">GMP.45</th>
<th data-quarto-table-cell-role="th">GMP.46</th>
<th data-quarto-table-cell-role="th">GMP.47</th>
<th data-quarto-table-cell-role="th">GMP.48</th>
<th data-quarto-table-cell-role="th">GMP.49</th>
<th data-quarto-table-cell-role="th">GMP.50</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Elane</td>
<td>0.050222</td>
<td>0.427325</td>
<td>-0.175609</td>
<td>-1.016725</td>
<td>-0.321307</td>
<td>-1.016725</td>
<td>-0.594497</td>
<td>-0.914855</td>
<td>-0.385123</td>
<td>-1.016725</td>
<td>...</td>
<td>1.829017</td>
<td>-0.724928</td>
<td>1.838109</td>
<td>1.043085</td>
<td>2.152079</td>
<td>0.224337</td>
<td>1.735849</td>
<td>0.988987</td>
<td>0.916954</td>
<td>1.764161</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Ctsg</td>
<td>-0.115593</td>
<td>0.195171</td>
<td>-0.306120</td>
<td>-0.167195</td>
<td>-0.185186</td>
<td>-0.429094</td>
<td>-0.761922</td>
<td>-0.810333</td>
<td>-0.336691</td>
<td>-0.787717</td>
<td>...</td>
<td>1.721915</td>
<td>-0.489532</td>
<td>1.469174</td>
<td>1.574845</td>
<td>1.852846</td>
<td>0.686645</td>
<td>1.290089</td>
<td>1.529435</td>
<td>1.273434</td>
<td>1.616981</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Mpo</td>
<td>-0.334801</td>
<td>0.195743</td>
<td>-0.409061</td>
<td>-0.412981</td>
<td>-0.223733</td>
<td>-0.240908</td>
<td>-0.836777</td>
<td>-0.948892</td>
<td>-0.401970</td>
<td>-1.104806</td>
<td>...</td>
<td>1.550774</td>
<td>-0.557452</td>
<td>1.422429</td>
<td>1.427961</td>
<td>1.859631</td>
<td>0.806051</td>
<td>1.454776</td>
<td>1.353039</td>
<td>1.098964</td>
<td>1.674420</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Car1</td>
<td>-0.501791</td>
<td>0.086734</td>
<td>-1.099513</td>
<td>-0.455777</td>
<td>-0.152378</td>
<td>0.208811</td>
<td>-0.928139</td>
<td>1.132393</td>
<td>-1.054360</td>
<td>-0.135432</td>
<td>...</td>
<td>-0.767363</td>
<td>0.269329</td>
<td>0.267178</td>
<td>-0.791656</td>
<td>-0.793902</td>
<td>-0.836976</td>
<td>-0.614000</td>
<td>-0.395873</td>
<td>-0.448685</td>
<td>-0.380978</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Ms4a3</td>
<td>-0.196420</td>
<td>0.016797</td>
<td>-0.522074</td>
<td>-0.806718</td>
<td>-0.806718</td>
<td>-0.806718</td>
<td>-0.611767</td>
<td>-0.600596</td>
<td>-0.568016</td>
<td>-0.509439</td>
<td>...</td>
<td>1.900349</td>
<td>-0.806718</td>
<td>1.421219</td>
<td>1.066604</td>
<td>2.000779</td>
<td>-0.090549</td>
<td>1.466089</td>
<td>1.363559</td>
<td>1.556162</td>
<td>1.782368</td>
</tr>
</tbody>
</table>

<p>5 rows × 150 columns</p>
</div>
</div>
</div>
<p>Lets make a boxplot of the data before and after scaling:</p>
<p>Before:</p>
<div id="4331634b" class="cell" data-execution_count="87">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>hspc_var.boxplot(rot<span class="op">=</span><span class="dv">90</span>)  <span class="co"># rot=90 rotates x-axis labels vertically (like las=2)</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-88-output-1.png" width="571" height="461" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>After:</p>
<div id="d18aea56" class="cell" data-execution_count="88">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>hspc_zs.boxplot(rot<span class="op">=</span><span class="dv">90</span>)  <span class="co"># rot=90 rotates x-axis labels</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-89-output-1.png" width="574" height="461" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<!-- ***Exercise***: Do some Googling and see if there is already a `zscore` function somewhere in Python that will calculate z-scores for you. Use it to z-score your data and save the output in `hspc_zs_v2`. There are maybe a couple of ways that you can do it. -->
</section>
<section id="clustering" class="level1">
<h1>Clustering</h1>
<p>All the steps before have been leading to this, clustering the data and making heatmaps - a staple method in bioinformatics.</p>
<p>Clustering is one of the most common visualisation techniques for genes expression data. Here we will learn how to do some basic histograms/heatmaps and plotting. To do this we need to bring in the <code>scipy</code> <a href="https://docs.scipy.org/doc/scipy/index.html">library</a> which is a library for engineering, maths, and stats. In this case we are going to pull in the <code>cluster</code> package and within, the <code>hierarchy</code> module.</p>
<div id="65725b07" class="cell" data-execution_count="89">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.cluster.hierarchy <span class="im">as</span> sch</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>linkage_matrix <span class="op">=</span> sch.linkage(hspc_zs, method<span class="op">=</span><span class="st">'ward'</span>,metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the dendrogram</span></span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a>sch.dendrogram(linkage_matrix, labels<span class="op">=</span>hspc_var.index.tolist(), leaf_rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Dendrogram of Rows (hspc_var)"</span>)</span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Sample"</span>)</span>
<span id="cb141-10"><a href="#cb141-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Distance"</span>)</span>
<span id="cb141-11"><a href="#cb141-11" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb141-12"><a href="#cb141-12" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-90-output-1.png" width="950" height="566" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="bce97f10" class="cell" data-execution_count="90">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.spatial.distance <span class="im">as</span> ssd</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="op">=</span> ssd.pdist(hspc_zs, metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Hierarchical clustering (like R's hclust)</span></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>linkage_matrix <span class="op">=</span> sch.linkage(dist_matrix, method<span class="op">=</span><span class="st">'ward'</span>)  <span class="co"># or "ward", "average", etc.</span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3: Plot dendrogram (like R's plot)</span></span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a>sch.dendrogram(linkage_matrix, labels<span class="op">=</span>hspc_zs.index.tolist(), leaf_rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Hierarchical Clustering of Genes"</span>)</span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Genes"</span>)</span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Distance"</span>)</span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a><span class="co">#lo = sch.dendrogram(linkage_matrix)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-91-output-1.png" width="950" height="566" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="a8e391f7" class="cell" data-execution_count="91">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.spatial.distance <span class="im">as</span> ssd</span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="op">=</span> ssd.pdist(hspc_var, metric<span class="op">=</span><span class="st">'correlation'</span>)</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Hierarchical clustering (like R's hclust)</span></span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a>linkage_matrix <span class="op">=</span> sch.linkage(dist_matrix, method<span class="op">=</span><span class="st">'ward'</span>)  <span class="co"># or "ward", "average", etc.</span></span>
<span id="cb143-6"><a href="#cb143-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb143-7"><a href="#cb143-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3: Plot dendrogram (like R's plot)</span></span>
<span id="cb143-8"><a href="#cb143-8" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb143-9"><a href="#cb143-9" aria-hidden="true" tabindex="-1"></a>sch.dendrogram(linkage_matrix, labels<span class="op">=</span>hspc_zs.index.tolist(), leaf_rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb143-10"><a href="#cb143-10" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Hierarchical Clustering of Genes"</span>)</span>
<span id="cb143-11"><a href="#cb143-11" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Genes"</span>)</span>
<span id="cb143-12"><a href="#cb143-12" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Distance"</span>)</span>
<span id="cb143-13"><a href="#cb143-13" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb143-14"><a href="#cb143-14" aria-hidden="true" tabindex="-1"></a>plt.show()</span>
<span id="cb143-15"><a href="#cb143-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb143-16"><a href="#cb143-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb143-17"><a href="#cb143-17" aria-hidden="true" tabindex="-1"></a><span class="co">#lo = sch.dendrogram(linkage_matrix)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-92-output-1.png" width="950" height="566" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Let say we want to cluster the samples(columns) instead? What should we do?</p>
<div id="a8b3fef3" class="cell" data-execution_count="92">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.cluster.hierarchy <span class="im">as</span> sch</span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>sample_linkage_matrix <span class="op">=</span> sch.linkage(hspc_zs.T, method<span class="op">=</span><span class="st">'ward'</span>,metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the dendrogram</span></span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))</span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a>sch.dendrogram(sample_linkage_matrix, labels<span class="op">=</span>hspc_var.T.index.tolist(), leaf_rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb144-8"><a href="#cb144-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"Dendrogram of Samples (hspc_var)"</span>)</span>
<span id="cb144-9"><a href="#cb144-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Sample"</span>)</span>
<span id="cb144-10"><a href="#cb144-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Distance"</span>)</span>
<span id="cb144-11"><a href="#cb144-11" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb144-12"><a href="#cb144-12" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-93-output-1.png" width="949" height="565" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="heatmaps" class="level2">
<h2 class="anchored" data-anchor-id="heatmaps">Heatmaps</h2>
<p>Heatmaps have been the go-to method for visualising gene expression data. We can make one from first principles using the work we have done above. Lets take a look at the contents of what is returned by <code>sch.dendrogram</code>.</p>
<p>Lets run these lines again, but this time saving the output from each call to cluster the data:</p>
<div id="9d40cfbb" class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>gene_linkage_matrix <span class="op">=</span> sch.linkage(hspc_zs, method<span class="op">=</span><span class="st">'ward'</span>,metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>genes_hc <span class="op">=</span> sch.dendrogram(gene_linkage_matrix, labels<span class="op">=</span>hspc_var.index.tolist(),no_plot<span class="op">=</span><span class="va">True</span>)<span class="co">#, leaf_rotation=90)</span></span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>sample_linkage_matrix <span class="op">=</span> sch.linkage(hspc_zs.T, method<span class="op">=</span><span class="st">'ward'</span>,metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a>samples_hc <span class="op">=</span> sch.dendrogram(sample_linkage_matrix, labels<span class="op">=</span>hspc_var.T.index.tolist(),no_plot<span class="op">=</span><span class="va">True</span>)<span class="co">#, leaf_rotation=90)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>*<strong>Exercise</strong>: What kind of object is <code>genes_hc</code>? Check it, and then get a list of what <code>genes_hc</code> contains.</p>
<div id="3a970023" class="cell" data-execution_count="94">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="bu">type</span>(genes_hc)</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>genes_hc.keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="90">
<pre><code>dict_keys(['icoord', 'dcoord', 'ivl', 'leaves', 'color_list', 'leaves_color_list'])</code></pre>
</div>
</div>
<p>At th</p>
<div id="d623d884" class="cell" data-execution_count="95">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>plt.imshow(hspc_zs, aspect<span class="op">=</span><span class="st">'auto'</span>, cmap<span class="op">=</span><span class="st">'viridis'</span>)</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-96-output-1.png" width="575" height="416" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can how ordered the rows according to <code>ivl</code></p>
<div id="7dff1923" class="cell" data-execution_count="96">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>hspc_ordered <span class="op">=</span> hspc_zs.loc[genes_hc[<span class="st">"ivl"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Replot the heatmap:</p>
<div id="e1a44ca8" class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>plt.imshow(hspc_ordered, aspect<span class="op">=</span><span class="st">'auto'</span>, cmap<span class="op">=</span><span class="st">'viridis'</span>)</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-98-output-1.png" width="575" height="416" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Looking better right? One more step and we will have clustered columns:</p>
<div id="aba13e0c" class="cell" data-execution_count="98">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>hspc_ordered <span class="op">=</span> hspc_zs.loc[genes_hc[<span class="st">"ivl"</span>],samples_hc[<span class="st">"ivl"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And plot again:</p>
<div id="98f0ed36" class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>plt.imshow(hspc_ordered, aspect<span class="op">=</span><span class="st">'auto'</span>, cmap<span class="op">=</span><span class="st">'viridis'</span>)</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-100-output-1.png" width="575" height="416" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>To make a heatap using <code>seaborn</code>:</p>
<div id="05b65bfc" class="cell" data-execution_count="100">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>sns.heatmap(hspc_ordered, cmap<span class="op">=</span><span class="st">'magma'</span>)  <span class="co"># Optional: cbar=True, xticklabels=False, yticklabels=False</span></span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Heatmap of Data'</span>)</span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-101-output-1.png" width="579" height="481" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can do a lot of this in one shot using the <code>clustermap function</code> where the clustering and plotting is done in one go:</p>
<div id="08caec39" class="cell" data-execution_count="101">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>sns.clustermap(hspc_var, method<span class="op">=</span><span class="st">'ward'</span>, metric<span class="op">=</span><span class="st">'euclidean'</span>, cmap<span class="op">=</span><span class="st">'magma'</span>, z_score<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-102-output-2.png" width="950" height="951" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>To save the plot do:</p>
<div id="e3cfb298" class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>hspc_clus <span class="op">=</span> sns.clustermap(hspc_var, method<span class="op">=</span><span class="st">'ward'</span>, metric<span class="op">=</span><span class="st">'euclidean'</span>, cmap<span class="op">=</span><span class="st">'magma'</span>, z_score<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>hspc_clus.fig.savefig(<span class="st">"clustermap.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-103-output-2.png" width="950" height="951" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can see from the heatmap that he genes <em>broadly</em> fall into 5 clusters, and it would be good to know which genes are in which cluster. To do this we need to probe the output of <code>hspc_clus</code> to get to that information.</p>
<div id="20a289f9" class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="bu">dir</span>(hspc_clus)</span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>gene_clusters_k5 <span class="op">=</span> sch.fcluster(hspc_clus.dendrogram_row.linkage,t<span class="op">=</span><span class="dv">5</span>, criterion<span class="op">=</span><span class="st">'maxclust'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have an array of cluster assignments, we can add this as a colour bar:</p>
<div id="d10bb056" class="cell" data-execution_count="104">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>row_colors <span class="op">=</span> pd.Series(gene_clusters_k5, index<span class="op">=</span>hspc_var.index, name<span class="op">=</span><span class="st">"cluster"</span>)</span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>unique_clusters <span class="op">=</span> <span class="bu">sorted</span>(row_colors.unique())</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a>palette <span class="op">=</span> sns.color_palette(<span class="st">"tab10"</span>, n_colors<span class="op">=</span><span class="bu">len</span>(unique_clusters))</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>lut <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(unique_clusters, palette))</span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a>row_colors_mapped <span class="op">=</span> row_colors.<span class="bu">map</span>(lut)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="bb0b4ca1" class="cell" data-execution_count="105">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>sns.clustermap(</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>    hspc_var,</span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a>    method<span class="op">=</span><span class="st">'ward'</span>,</span>
<span id="cb160-4"><a href="#cb160-4" aria-hidden="true" tabindex="-1"></a>    metric<span class="op">=</span><span class="st">'euclidean'</span>,</span>
<span id="cb160-5"><a href="#cb160-5" aria-hidden="true" tabindex="-1"></a>    cmap<span class="op">=</span><span class="st">'magma'</span>,</span>
<span id="cb160-6"><a href="#cb160-6" aria-hidden="true" tabindex="-1"></a>    z_score<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb160-7"><a href="#cb160-7" aria-hidden="true" tabindex="-1"></a>    row_colors<span class="op">=</span>row_colors_mapped</span>
<span id="cb160-8"><a href="#cb160-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb160-9"><a href="#cb160-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-106-output-2.png" width="950" height="951" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can also count up how many genes we have in each cluster:</p>
<div id="40412a9e" class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>counts<span class="op">=</span> pd.Series(gene_clusters_k5).value_counts().sort_index()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="12ee6e16" class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>counts.plot(kind<span class="op">=</span><span class="st">'bar'</span>)</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-108-output-1.png" width="575" height="407" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="linear-models" class="level1">
<h1>Linear models</h1>
<p>Fitting a linear model (or linear regression) is also another fairly common thing to do, and actually forms the basis of some of the more famous packages we use such as DESeq2, limma etc. Let’s use our expression data to demonstrate this.</p>
<p>In our case we have 3 different samples:</p>
<div id="fde0f82d" class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>hspc_data.columns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="104">
<pre><code>Index(['LTHSC.1', 'LTHSC.2', 'LTHSC.3', 'LTHSC.4', 'LTHSC.5', 'LTHSC.6',
       'LTHSC.7', 'LTHSC.8', 'LTHSC.9', 'LTHSC.10',
       ...
       'GMP.41', 'GMP.42', 'GMP.43', 'GMP.44', 'GMP.45', 'GMP.46', 'GMP.47',
       'GMP.48', 'GMP.49', 'GMP.50'],
      dtype='object', length=150)</code></pre>
</div>
</div>
<p>But what we need to do is make a vector where the <code>.X</code> has been clipped off to make a vector of three groups. We do this by:</p>
<div id="4534999f" class="cell" data-execution_count="109">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>groups <span class="op">=</span> pd.Series(hspc_data.columns).<span class="bu">str</span>.replace(<span class="vs">r"</span><span class="ch">\.</span><span class="dv">.</span><span class="op">*</span><span class="vs">"</span>, <span class="st">""</span>, regex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(groups.tolist())  <span class="co"># ['stem', 'stem', 'prog', 'prog']</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'LTHSC', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'MEP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP', 'GMP']</code></pre>
</div>
</div>
<p>We can now make a linear model for the first gene from this array of groups:</p>
<div id="719f74ff" class="cell" data-execution_count="110">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.api <span class="im">as</span> sm</span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.formula.api <span class="im">as</span> smf</span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">'expression'</span>: hspc_data.iloc[<span class="dv">0</span>].values,</span>
<span id="cb168-6"><a href="#cb168-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">'group'</span>: groups.values</span>
<span id="cb168-7"><a href="#cb168-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb168-8"><a href="#cb168-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb168-9"><a href="#cb168-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Fit the model</span></span>
<span id="cb168-10"><a href="#cb168-10" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> smf.ols(<span class="st">'expression ~ group'</span>, data<span class="op">=</span>df).fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>…and run the ANOVA:</p>
<div id="e55539c1" class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>anova_table <span class="op">=</span> sm.stats.anova_lm(model, typ<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(anova_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               sum_sq     df         F    PR(&gt;F)
group       26.442809    2.0  1.500594  0.226393
Residual  1295.184848  147.0       NaN       NaN</code></pre>
</div>
</div>
<div id="b17ade0c" class="cell" data-execution_count="112">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>p_value <span class="op">=</span> anova_table[<span class="st">"PR(&gt;F)"</span>][<span class="st">"group"</span>]</span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(p_value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.22639342020278108</code></pre>
</div>
</div>
<p><strong><em>Exercise</em></strong>: Make a function called do_anova which takes a vector vand a vector groups, performs linear modelling and ANOVA and returns just the p-value.</p>
<div id="fda210a3" class="cell" data-execution_count="113">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> do_anova(vals,grps):</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">'expression'</span>: vals,</span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">'group'</span>: grps</span>
<span id="cb173-5"><a href="#cb173-5" aria-hidden="true" tabindex="-1"></a>        })</span>
<span id="cb173-6"><a href="#cb173-6" aria-hidden="true" tabindex="-1"></a>    mod <span class="op">=</span> smf.ols(<span class="st">'expression ~ group'</span>, data<span class="op">=</span>df).fit()</span>
<span id="cb173-7"><a href="#cb173-7" aria-hidden="true" tabindex="-1"></a>    anova_table <span class="op">=</span> sm.stats.anova_lm(mod, typ<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb173-8"><a href="#cb173-8" aria-hidden="true" tabindex="-1"></a>    p_value <span class="op">=</span> anova_table[<span class="st">"PR(&gt;F)"</span>][<span class="st">"group"</span>]</span>
<span id="cb173-9"><a href="#cb173-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> p_value</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="1fafee12" class="cell" data-execution_count="114">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>do_anova(hspc_data.iloc[<span class="dv">0</span>].values,groups)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="110">
<pre><code>np.float64(0.22639342020278108)</code></pre>
</div>
</div>
<p>We can now apply this to all rows of our data in one shot!</p>
<div id="9de6216f" class="cell" data-execution_count="115">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>hspc_pvals <span class="op">=</span> hspc_data.<span class="bu">apply</span>(<span class="kw">lambda</span> row: do_anova(row.values, groups), axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We need to correct the p-values! We usually do this using Benjamini-Hochberg correction:</p>
<div id="d022bb06" class="cell" data-execution_count="116">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> ss</span>
<span id="cb177-2"><a href="#cb177-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-3"><a href="#cb177-3" aria-hidden="true" tabindex="-1"></a>hspc_pcor <span class="op">=</span> ss.false_discovery_control(hspc_pvals)</span>
<span id="cb177-4"><a href="#cb177-4" aria-hidden="true" tabindex="-1"></a>p_cors_df <span class="op">=</span> pd.DataFrame({</span>
<span id="cb177-5"><a href="#cb177-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">'pvals'</span>: hspc_pcor,</span>
<span id="cb177-6"><a href="#cb177-6" aria-hidden="true" tabindex="-1"></a>        })</span>
<span id="cb177-7"><a href="#cb177-7" aria-hidden="true" tabindex="-1"></a>p_cors_df.index <span class="op">=</span> hspc_data.index</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can pull out the significant genes using a conditional:</p>
<div id="6a0458df" class="cell" data-execution_count="117">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>sig_genes <span class="op">=</span> p_cors_df.index[p_cors_df[<span class="st">'pvals'</span>] <span class="op">&lt;</span> <span class="fl">0.001</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Get the genes from the main matrix:</p>
<div id="2faa1ff2" class="cell" data-execution_count="118">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>sig_data <span class="op">=</span> hspc_data.loc[sig_genes]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And heatmap them!</p>
<div id="390ec16d" class="cell" data-execution_count="119">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a>sns.clustermap(</span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a>    sig_data,</span>
<span id="cb180-3"><a href="#cb180-3" aria-hidden="true" tabindex="-1"></a>    method<span class="op">=</span><span class="st">'ward'</span>,</span>
<span id="cb180-4"><a href="#cb180-4" aria-hidden="true" tabindex="-1"></a>    metric<span class="op">=</span><span class="st">'euclidean'</span>,</span>
<span id="cb180-5"><a href="#cb180-5" aria-hidden="true" tabindex="-1"></a>    cmap<span class="op">=</span><span class="st">'magma'</span>,</span>
<span id="cb180-6"><a href="#cb180-6" aria-hidden="true" tabindex="-1"></a>    z_score<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb180-7"><a href="#cb180-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
/home/shamit/anaconda3/envs/quarto.env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-120-output-2.png" width="950" height="951" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="kmeans" class="level1">
<h1>Kmeans</h1>
<p>We’re going to program our first algorithm together! Kmeans clustering is a common way to cluster expression data, and the algorithm is actually pretty simple, and a great way to learn how to think programmatically. The data we’ll use comes from yeast where the expression of 256 genes have been measured as a synchronised population go through two divisions. It’s a nice small dataset with clear clusters.</p>
<p>There is a file called <a href="https://github.com/shambam/R_programming_1/blob/main/Spellman_Yeast_Cell_Cycle.tsv">Spellman_Yeast_Cell_Cycle.tsv</a>. Load it into an object called <code>ycc</code>.</p>
<div id="e1580f16" class="cell" data-execution_count="120">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a>ycc <span class="op">=</span> pd.read_csv(<span class="st">"Spellman_Yeast_Cell_Cycle.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, header<span class="op">=</span><span class="dv">0</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb182-2"><a href="#cb182-2" aria-hidden="true" tabindex="-1"></a>ycc_np <span class="op">=</span> ycc.values</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="bcbe5c13" class="cell" data-execution_count="121">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a>ran <span class="op">=</span> np.random.choice(np.arange(<span class="dv">6</span>), size<span class="op">=</span><span class="dv">150</span>, replace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data has already been z-scored, so we need to learn how to calculate the Euclidean distance between 2 genes (<span class="math inline">\(g^{i}\)</span> and <span class="math inline">\(g^{j}\)</span>) (vector). We do this using the formula:</p>
<p><span class="math display">\[d_{ij}=\sqrt{\sum{(g^{i}_t-g^{j}_t)^2}}\]</span></p>
<p>Write a function called <code>e_dist</code> which takes two vectors <code>v1</code> and <code>v2</code> and calculates the distance between them. In our data, these two vectors (<code>v1</code> and <code>v2</code>) are the different expression values of two genes through the different time points measured in the yeast cell cycle .</p>
<div id="8c077e44" class="cell" data-execution_count="122">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> math</span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> e_dist(v1,v2):</span>
<span id="cb184-4"><a href="#cb184-4" aria-hidden="true" tabindex="-1"></a>    d <span class="op">=</span> math.sqrt(<span class="bu">pow</span>(<span class="bu">sum</span>((v1<span class="op">-</span>v2)),<span class="dv">2</span>))</span>
<span id="cb184-5"><a href="#cb184-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> d</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The kmeans algorithm needs the user to say how many clusters are being searched for, so in this case we’ll say <strong>8</strong>.</p>
<ol type="1">
<li>Pick 8 genes (rows) at random to act as our initial centroids.</li>
<li>Calculate the distance between each gene to each of the 8 centroids.</li>
<li>Assign each gene to its closest centroid.</li>
<li>For each of the 8 clusters, calculate a new centroid.</li>
<li>Repeat 100 times steps 2, 3, and 4.</li>
<li>We’ll now work through this problem together and crowd source a solution.</li>
</ol>
<div id="209b00df" class="cell" data-execution_count="123">
<details class="code-fold">
<summary>Reveal solution</summary>
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>K<span class="op">=</span><span class="dv">8</span></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a><span class="bu">iter</span><span class="op">=</span><span class="dv">100</span></span>
<span id="cb185-3"><a href="#cb185-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-4"><a href="#cb185-4" aria-hidden="true" tabindex="-1"></a>ran <span class="op">=</span> np.random.choice(np.arange(ycc_np.shape[<span class="dv">0</span>]), size<span class="op">=</span><span class="dv">8</span>, replace<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb185-5"><a href="#cb185-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(ran)</span>
<span id="cb185-6"><a href="#cb185-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-7"><a href="#cb185-7" aria-hidden="true" tabindex="-1"></a>centroids <span class="op">=</span> ycc_np[ran,:]</span>
<span id="cb185-8"><a href="#cb185-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-9"><a href="#cb185-9" aria-hidden="true" tabindex="-1"></a>centroids</span>
<span id="cb185-10"><a href="#cb185-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-11"><a href="#cb185-11" aria-hidden="true" tabindex="-1"></a>clusters<span class="op">=</span><span class="va">None</span></span>
<span id="cb185-12"><a href="#cb185-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-13"><a href="#cb185-13" aria-hidden="true" tabindex="-1"></a>dists <span class="op">=</span> np.zeros((ycc_np.shape[<span class="dv">0</span>],K))</span>
<span id="cb185-14"><a href="#cb185-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-15"><a href="#cb185-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">iter</span>):</span>
<span id="cb185-16"><a href="#cb185-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-17"><a href="#cb185-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> gene <span class="kw">in</span> <span class="bu">range</span>(ycc_np.shape[<span class="dv">0</span>]):</span>
<span id="cb185-18"><a href="#cb185-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-19"><a href="#cb185-19" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(K):</span>
<span id="cb185-20"><a href="#cb185-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-21"><a href="#cb185-21" aria-hidden="true" tabindex="-1"></a>            dists[gene,k] <span class="op">=</span> e_dist(ycc_np[gene,:],centroids[k,:])</span>
<span id="cb185-22"><a href="#cb185-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-23"><a href="#cb185-23" aria-hidden="true" tabindex="-1"></a>    clusters <span class="op">=</span> np.argmin(dists, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb185-24"><a href="#cb185-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb185-25"><a href="#cb185-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> nc <span class="kw">in</span> <span class="bu">range</span>(K):</span>
<span id="cb185-26"><a href="#cb185-26" aria-hidden="true" tabindex="-1"></a>        centroids[nc,:] <span class="op">=</span> np.mean(ycc_np[clusters<span class="op">==</span>nc,:],axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[255 146  93 105 213 205 212 204]</code></pre>
</div>
</div>
<div id="b15ec098" class="cell" data-execution_count="124">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">4</span>, figsize<span class="op">=</span>(<span class="dv">16</span>, <span class="dv">8</span>))</span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a>axes <span class="op">=</span> axes.flatten()</span>
<span id="cb187-3"><a href="#cb187-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-4"><a href="#cb187-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">8</span>):</span>
<span id="cb187-5"><a href="#cb187-5" aria-hidden="true" tabindex="-1"></a>    ax <span class="op">=</span> axes[k]</span>
<span id="cb187-6"><a href="#cb187-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb187-7"><a href="#cb187-7" aria-hidden="true" tabindex="-1"></a>    ycc_c <span class="op">=</span> ycc_np[clusters <span class="op">==</span> k, :]  <span class="co"># rows in cluster k</span></span>
<span id="cb187-8"><a href="#cb187-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> ycc_c.shape[<span class="dv">0</span>] <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb187-9"><a href="#cb187-9" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span>  <span class="co"># skip if no rows</span></span>
<span id="cb187-10"><a href="#cb187-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb187-11"><a href="#cb187-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot the first row</span></span>
<span id="cb187-12"><a href="#cb187-12" aria-hidden="true" tabindex="-1"></a>    ax.plot(ycc_c[<span class="dv">0</span>], color<span class="op">=</span><span class="st">'black'</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb187-13"><a href="#cb187-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb187-14"><a href="#cb187-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Overlay the rest</span></span>
<span id="cb187-15"><a href="#cb187-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(ycc_c.shape[<span class="dv">0</span>]):</span>
<span id="cb187-16"><a href="#cb187-16" aria-hidden="true" tabindex="-1"></a>        ax.plot(ycc_c[i], color<span class="op">=</span><span class="st">'black'</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb187-17"><a href="#cb187-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-18"><a href="#cb187-18" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="ss">f"Cluster </span><span class="sc">{</span>k<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb187-19"><a href="#cb187-19" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(ycc_np.<span class="bu">min</span>(), ycc_np.<span class="bu">max</span>())</span>
<span id="cb187-20"><a href="#cb187-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb187-21"><a href="#cb187-21" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb187-22"><a href="#cb187-22" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-125-output-1.png" width="1526" height="757" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="d13d4074" class="cell" data-execution_count="125">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assume ycc and clusters already exist</span></span>
<span id="cb188-2"><a href="#cb188-2" aria-hidden="true" tabindex="-1"></a>n_samples, n_timepoints <span class="op">=</span> ycc.shape</span>
<span id="cb188-3"><a href="#cb188-3" aria-hidden="true" tabindex="-1"></a>timepoints <span class="op">=</span> np.arange(n_timepoints)</span>
<span id="cb188-4"><a href="#cb188-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-5"><a href="#cb188-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a DataFrame from ycc</span></span>
<span id="cb188-6"><a href="#cb188-6" aria-hidden="true" tabindex="-1"></a>ycc_df <span class="op">=</span> pd.DataFrame(ycc)</span>
<span id="cb188-7"><a href="#cb188-7" aria-hidden="true" tabindex="-1"></a>ycc_df[<span class="st">'cluster'</span>] <span class="op">=</span> clusters</span>
<span id="cb188-8"><a href="#cb188-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-9"><a href="#cb188-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Melt the DataFrame to long format</span></span>
<span id="cb188-10"><a href="#cb188-10" aria-hidden="true" tabindex="-1"></a>long_df <span class="op">=</span> ycc_df.reset_index().melt(id_vars<span class="op">=</span>[<span class="st">'index'</span>, <span class="st">'cluster'</span>], </span>
<span id="cb188-11"><a href="#cb188-11" aria-hidden="true" tabindex="-1"></a>                                    var_name<span class="op">=</span><span class="st">'time'</span>, value_name<span class="op">=</span><span class="st">'value'</span>)</span>
<span id="cb188-12"><a href="#cb188-12" aria-hidden="true" tabindex="-1"></a>long_df.rename(columns<span class="op">=</span>{<span class="st">'index'</span>: <span class="st">'sample'</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb188-13"><a href="#cb188-13" aria-hidden="true" tabindex="-1"></a>long_df[<span class="st">'time'</span>] <span class="op">=</span> long_df[<span class="st">'time'</span>].<span class="bu">str</span>.extract(<span class="vs">r'</span><span class="kw">(</span><span class="dv">\d</span><span class="op">+</span><span class="kw">)</span><span class="vs">'</span>).astype(<span class="bu">int</span>) <span class="co"># make sure time is numeric</span></span>
<span id="cb188-14"><a href="#cb188-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-15"><a href="#cb188-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with seaborn FacetGrid</span></span>
<span id="cb188-16"><a href="#cb188-16" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> sns.FacetGrid(long_df, col<span class="op">=</span><span class="st">'cluster'</span>, col_wrap<span class="op">=</span><span class="dv">4</span>, sharey<span class="op">=</span><span class="va">True</span>, height<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb188-17"><a href="#cb188-17" aria-hidden="true" tabindex="-1"></a>g.map_dataframe(sns.lineplot, x<span class="op">=</span><span class="st">'time'</span>, y<span class="op">=</span><span class="st">'value'</span>, alpha<span class="op">=</span><span class="fl">0.3</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb188-18"><a href="#cb188-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb188-19"><a href="#cb188-19" aria-hidden="true" tabindex="-1"></a>g.set_titles(<span class="st">"Cluster </span><span class="sc">{col_name}</span><span class="st">"</span>)</span>
<span id="cb188-20"><a href="#cb188-20" aria-hidden="true" tabindex="-1"></a>g.set_axis_labels(<span class="st">"Time"</span>, <span class="st">"Value"</span>)</span>
<span id="cb188-21"><a href="#cb188-21" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb188-22"><a href="#cb188-22" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cell-126-output-1.png" width="1141" height="563" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="the-final-project" class="level1">
<h1>The Final Project</h1>
<p>Your final project is going to combine everything you have learned over the last few days to tackle a final project you will do in your own time over the next X weeks. It will be along the same lines the Kmeans clustering we just did, but this time we’re going to cluster the data using the Metropolis-Hastings algorithm. This is a simulated annealing process, so we need to learn how that works first.</p>
<section id="optimisation-problems" class="level2">
<h2 class="anchored" data-anchor-id="optimisation-problems">Optimisation problems</h2>
<p>Programming an optimisation problem is a great way of learning how to code. In this case we are going to tackle a minimisation problem. We can think of well clustered data having low energy, in that, each cluster is tight and has little within cluster variance. If we calculate the variance <em>within</em> each cluster, and sum over all clusters, we get the total variance (energy) of the system. Lets go back to the equation we know that measures the distance between two genes <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> over <span class="math inline">\(t\)</span> timepoints:</p>
<p><span class="math display">\[d_{ij}=\sqrt{\sum{(g^{i}_t-g^{j}_t)^2}}\]</span></p>
<p>To measure the compactness of a clustering, we sum the pairwise distances for each cluster <span class="math inline">\(K\)</span>, then sum over all <span class="math inline">\(K\)</span>s and them divide by <span class="math inline">\(K\)</span>.</p>
<p><span class="math display">\[ E(K)=\frac{1}{K}\sum^K_{k=1} \left[ \sum_{i\epsilon Ck}\sum_{j\epsilon Ck} d_{ij}\right] \]</span></p>
<p>For a well clustered data, <span class="math inline">\(E(K)\)</span> should be as <strong>small</strong> as possible. Lets say we have 1000 genes, and we want to partition them into 10 clusters. The number of combinations is too high for us to try each one to brute force a true <span class="math inline">\(E\)</span>. This is why we use a <em>heuristic</em> algorithm to get us as close to the solution as possible in a smaller amount of time.</p>
<p>If we tried to visualise the energy landscape we can imagine it might look something like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/EnergyLandscape.png" class="img-fluid figure-img" width="600"></p>
<figcaption>Energy</figcaption>
</figure>
</div>
<p><strong>The first thing you need to do is rescale each gene (row) of the data so the smallest value is given 0, and the highest value is given 1.</strong></p>
<p>Lets go through the process of the algorithm. To do this you need 4 parameters:</p>
<ol type="1">
<li>The temperature of the system <span class="math inline">\(Temp\)</span></li>
<li>Cooling factor <span class="math inline">\(cool\)</span></li>
<li>Number of clusters <span class="math inline">\(K\)</span></li>
<li>How many iterations to perform <span class="math inline">\(I\)</span>.</li>
</ol>
<p>Algorithm goes:</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Randomly assign each gene to one of your K clusters.</p>
<p>for each iteration <span class="math inline">\(I\)</span> {</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Calculate <span class="math inline">\(E\)</span>. Call this <span class="math inline">\(E_{old}\)</span></p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Randomly select a single gene and assign it randomly to another cluster.</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Calculate <span class="math inline">\(E_{new}\)</span></p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if(<span class="math inline">\(E_{new}\)</span> &lt; <span class="math inline">\(E_{old}\)</span>) {accept the move}</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if(<span class="math inline">\(E_{new} &gt; E_{old})\)</span> {if(<span class="math inline">\(e^{-\frac{E_{new}-E_{old}}{Temp}} &gt; R(0,1)\)</span>) {accept the move} }</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;else{reject the move}</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;Set a new <span class="math inline">\(Temp\)</span> by doing <span class="math inline">\(Temp=Temp \times cool\)</span> }</p>
<p>Done</p>
<p><span class="math inline">\(R(0,1)\)</span> is a randomly generated number from a uniform distribution that lies between 0 and 1.</p>
<p><strong><em>Things to do</em></strong></p>
<ol type="1">
<li>Code this algorithm using the same yeast timecourse data.</li>
<li>Put this algorithm into a function.</li>
<li>When running the function, make sure the starting <span class="math inline">\(E_{start}\)</span> and final <span class="math inline">\(E_{final}\)</span> are printed to the screen.</li>
</ol>
<p><strong><em>Tip</em></strong></p>
<p>Try and break the problem up into functions which can be called when needed. It will be easier to write the main algorithm.</p>
<p><strong><em>Why this exercise is a good one</em></strong></p>
<p>By tackling a more difficult problem the idea is that you will dig deeper into the language and really understand how things work in the background of the functions you use in packages such as Seurat etc.</p>
<p><strong><em>Are there libraries in Python to do this?</em></strong></p>
<p>Yes there are, but you won’t learn anything by using them which is why we are coding this “by hand”.</p>
<p><strong><em>A little inspiration</em></strong></p>
<p>Before you start, read <a href="https://www.newyorker.com/magazine/2018/12/10/the-friendship-that-made-google-huge">this</a>, and then watch <a href="https://youtu.be/USC6c9Dtak8">this</a>.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    // Ensure there is a toggle, if there isn't float one in the top right
    if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
      const a = window.document.createElement('a');
      a.classList.add('top-right');
      a.classList.add('quarto-color-scheme-toggle');
      a.href = "";
      a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
      const i = window.document.createElement("i");
      i.classList.add('bi');
      a.appendChild(i);
      window.document.body.appendChild(a);
    }
    setColorSchemeToggle(hasAlternateSentinel())
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>